crates/bevy_render/src/texture/dds.rs

```diff
@@ -277,3 +276,95 @@ pub fn dds_format_to_texture_format(
         ));
     })
 }
+
+#[cfg(test)]
+mod test {
+    use wgpu::{util::TextureDataOrder, TextureDescriptor, TextureDimension};
+
+    use crate::texture::CompressedImageFormats;
+
+    use super::dds_buffer_to_image;
+
+    /// `wgpu::create_texture_with_data` that reads from data structure but doesn't actually talk to your GPU
+    fn fake_wgpu_create_texture_with_data(desc: &TextureDescriptor<'_>, data: &[u8]) {
+        // Will return None only if it's a combined depth-stencil format
+        // If so, default to 4, validation will fail later anyway since the depth or stencil
+        // aspect needs to be written to individually
+        let block_size = desc.format.block_copy_size(None).unwrap_or(4);
+        let (block_width, block_height) = desc.format.block_dimensions();
+        let layer_iterations = desc.array_layer_count();
+
+        let outer_iteration;
+        let inner_iteration;
+        match TextureDataOrder::default() {
+            TextureDataOrder::LayerMajor => {
+                outer_iteration = layer_iterations;
+                inner_iteration = desc.mip_level_count;
+            }
+            TextureDataOrder::MipMajor => {
+                outer_iteration = desc.mip_level_count;
+                inner_iteration = layer_iterations;
+            }
+        }
+
+        let mut binary_offset = 0;
+        for outer in 0..outer_iteration {
+            for inner in 0..inner_iteration {
+                let (_layer, mip) = match TextureDataOrder::default() {
+                    TextureDataOrder::LayerMajor => (outer, inner),
+                    TextureDataOrder::MipMajor => (inner, outer),
+                };
+
+                let mut mip_size = desc.mip_level_size(mip).unwrap();
+                // copying layers separately
+                if desc.dimension != TextureDimension::D3 {
+                    mip_size.depth_or_array_layers = 1;
+                }
+
+                // When uploading mips of compressed textures and the mip is supposed to be
+                // a size that isn't a multiple of the block size, the mip needs to be uploaded
+                // as its "physical size" which is the size rounded up to the nearest block size.
+                let mip_physical = mip_size.physical_size(desc.format);
+
+                // All these calculations are performed on the physical size as that's the
+                // data that exists in the buffer.
+                let width_blocks = mip_physical.width / block_width;
+                let height_blocks = mip_physical.height / block_height;
+
+                let bytes_per_row = width_blocks * block_size;
+                let data_size = bytes_per_row * height_blocks * mip_size.depth_or_array_layers;
+
+                let end_offset = binary_offset + data_size as usize;
+
+                assert!(binary_offset < data.len());
+                assert!(end_offset <= data.len());
+                // data[binary_offset..end_offset])
```

Comment:
Debug?

Reply:
Not exactly. The commented line is how wgpu reads the data. I wanted to leave it here as a comment to show why the asserts are there.